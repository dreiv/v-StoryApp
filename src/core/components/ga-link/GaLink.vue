<script setup lang="ts">
import { useCssModule, computed, type AnchorHTMLAttributes, useAttrs } from 'vue'

export interface GaLinkProps extends /* @vue-ignore */ AnchorHTMLAttributes {
  text?: string
  size?: 'small' | 'medium' | 'large'
  inline?: boolean
  disabled?: boolean
}

defineOptions({ inheritAttrs: false })

const { size = 'medium', inline, disabled } = defineProps<GaLinkProps>()
const style = useCssModule()
const attrs = useAttrs()

const classes = computed(() => [
  style.link,
  size && style[size],
  inline && style.inline,
  disabled && style.disabled,
])

const attributes = computed(() => ({
  ...attrs,
  href: disabled ? undefined : (attrs?.href as string | undefined),
  'aria-disabled': disabled ? true : undefined,
}))
</script>

<template>
  <a :class="classes" v-bind="attributes">
    <slot>{{ text }}</slot>
  </a>
</template>

<style module>
.link {
  display: inline-flex;
  align-items: center;
  gap: var(--ga-size-spacing-02);

  cursor: pointer;
  border-radius: var(--ga-radius);
  padding-inline: var(--ga-size-spacing-01);

  color: var(--ga-color-text-action);
  font-weight: var(--ga-font-weight-medium);
  font-size: var(--ga-text-md-font-size);
  line-height: var(--ga-text-md-line-height);
  text-decoration: inherit;

  &.disabled {
    cursor: not-allowed;
    color: var(--ga-color-text-disabled);

    & .link_icon {
      color: var(--ga-color-icon-disabled);
    }
  }

  &:not(.disabled) {
    &:visited {
      color: var(--ga-color-text-link-visited);
      text-decoration-color: var(--ga-color-text-link-visited);
    }

    &:hover {
      color: var(--ga-color-text-action-hover);
      text-decoration: underline var(--ga-size-border-width-sm) solid var(--ga-color-primary-90);

      &.inline {
        text-decoration-thickness: var(--ga-size-border-width-md);
      }
    }

    &:active {
      color: var(--ga-color-text-action);
      text-decoration-color: var(--ga-color-text-action);
    }

    &:focus-visible {
      outline: var(--ga-size-border-width-md) solid var(--ga-color-border-focus);
      outline-offset: var(--ga-size-spacing-01);
    }
  }
}

.small {
  font-size: var(--ga-text-sm-font-size);
  line-height: var(--ga-text-sm-line-height);
}

.medium {
  font-size: var(--ga-text-md-font-size);
  line-height: var(--ga-text-md-line-height);
}

.large {
  font-size: var(--ga-text-lg-font-size);
  line-height: var(--ga-text-lg-line-height);
}

.inline {
  text-decoration: underline var(--ga-size-border-width-sm) solid var(--ga-color-primary-90);
}
</style>
