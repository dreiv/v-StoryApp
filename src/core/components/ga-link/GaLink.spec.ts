import { shallowMount, type VueWrapper } from '@vue/test-utils'
import GaLink from './GaLink.vue'

describe('GaLink', () => {
  let wrapper: VueWrapper

  it('should render correctly with default props', () => {
    wrapper = shallowMount(GaLink)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correctly with text prop', () => {
    wrapper = shallowMount(GaLink, { props: { text: 'Test Link' } })

    expect(wrapper.text()).toContain('Test Link')
  })

  it('should render correctly with default slot', () => {
    wrapper = shallowMount(GaLink, { slots: { default: 'Slot Link' } })

    expect(wrapper.text()).toContain('Slot Link')
  })

  it.each(['small', 'medium', 'large'])(
    'should apply the %s class when size prop is %s',
    async (size) => {
      wrapper = shallowMount(GaLink)
      await wrapper.setProps({ size })

      expect(wrapper.classes()).toContain(size)
    },
  )

  it('should apply the inline class when inline prop is true', async () => {
    wrapper = shallowMount(GaLink)
    await wrapper.setProps({ inline: true })

    expect(wrapper.classes()).toContain('inline')
  })

  it('should apply the disabled class and aria-disabled attribute when disabled prop is true', async () => {
    wrapper = shallowMount(GaLink)
    await wrapper.setProps({ disabled: true })

    expect(wrapper.classes()).toContain('disabled')
    expect(wrapper.attributes('aria-disabled')).toBe('true')
    expect(wrapper.attributes('href')).toBeUndefined()
  })

  it('should pass href attribute', () => {
    wrapper = shallowMount(GaLink, { attrs: { href: '#test' } })
    expect(wrapper.attributes('href')).toBe('#test')
  })
})
