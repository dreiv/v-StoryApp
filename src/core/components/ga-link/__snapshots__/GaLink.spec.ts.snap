// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GaLink > should render correctly with default props 1`] = `"<a class="link"></a>"`;
